modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/DeployService.java

```diff
@@ -202,10 +203,39 @@ public void asyncDeployService(Deployment deployment, DeployTask deployTask) {
             log.error("asyncDeployService failed.", e);
             deployServiceEntity.setServiceDeploymentState(ServiceDeploymentState.DEPLOY_FAILED);
             deployServiceEntity.setResultMessage(e.getMessage());
+            deployServiceEntity.setProperties(deployResult.getProperties());
+            deployServiceEntity.setPrivateProperties(deployResult.getPrivateProperties());
+            deployServiceEntity.setDeployResourceList(
+                    getDeployResourceEntityList(deployResult.getResources(), deployServiceEntity));
             maskSensitiveFields(deployServiceEntity);
             deployServiceStorage.storeAndFlush(deployServiceEntity);
+            rollbackOnDeploymentFailure(deployServiceEntity, deployResult, deployment, deployTask);
         }
+    }
 
+    /**
+     * Perform rollback when deployment fails and destroy the created resources.
+     */
+    private void rollbackOnDeploymentFailure(DeployServiceEntity deployServiceEntity,
+                                             DeployResult deployResult, Deployment deployment,
+                                             DeployTask deployTask) {
+        if (!CollectionUtils.isEmpty(deployResult.getResources())) {
+            String stateFile = deployServiceEntity.getPrivateProperties().get(STATE_FILE_NAME);
+            DeployResult destroyResult = deployment.destroy(deployTask, stateFile);
+            if (destroyResult.getState() == TerraformExecState.DESTROY_SUCCESS) {
+                deployServiceEntity.setProperties(destroyResult.getProperties());
+                deployServiceEntity.setPrivateProperties(destroyResult.getPrivateProperties());
+                deployServiceEntity.setDeployResourceList(
+                        getDeployResourceEntityList(destroyResult.getResources(),
+                                deployServiceEntity));
+            } else {
+                deployServiceEntity.setServiceDeploymentState(
+                        ServiceDeploymentState.MANUAL_CLEANUP_REQUIRED);
+            }
+            if (deployServiceStorage.storeAndFlush(deployServiceEntity)) {
+                deployment.deleteTaskWorkspace(deployTask.getId().toString());
+            }
+        }
```

Comment:
please a log line for else block -> "No resources to rollback. "

Reply:
fixed